projects/graph/src/draw.py

```diff
@@ -10,5 +12,107 @@
 
 class BokehGraph:
     """Class that takes a graph and exposes drawing methods."""
-    def __init__(self):
-        pass  # TODO
+    def __init__(self, graph, title='Graph', width=100, height=100,
+                 show_axis=False, show_grid=False, circle_size=35,
+                 draw_components=False):
+        if not graph.vertices:
+            raise Exception('Graph should contain vertices!')
+        self.graph = graph
+        self.width = width
+        self.height = height
+        self.pos = {}  # dict to map vertices to x, y positions
+        # Set up plot, the canvas/space to draw on
+        self.plot = figure(title=title, x_range=(0, width), y_range=(0, height))
+        self.plot.axis.visible = show_axis
+        self.plot.grid.visible = show_grid
+        self._setup_graph_renderer(circle_size, draw_components)
+        self._setup_labels()
+
+
+    def _setup_graph_renderer(self, circle_size, draw_components):
+        # The renderer will have the actual logic for drawing
+        graph_renderer = GraphRenderer()
+        # Saving vertices in an arbitrary but persistent order
+        self.vertex_list = list(self.graph.vertices.keys())
+
+        # Add the vertex data as instructions for drawing nodes
+        graph_renderer.node_renderer.data_source.add(
+            [vertex.label for vertex in self.vertex_list], 'index')
+        colors = (self._get_connected_component_colors() if draw_components
+                  else self._get_random_colors())
+        graph_renderer.node_renderer.data_source.add(colors, 'color')
+        # And circles
+        graph_renderer.node_renderer.glyph = Circle(size=circle_size,
+                                                    fill_color='color')
+
+        # Add the edge [start, end] indices as instructions for drawing edges
+        graph_renderer.edge_renderer.data_source.data = self._get_edge_indexes()
+        self.randomize()  # Randomize vertex coordinates, and set as layout
+        graph_renderer.layout_provider = StaticLayoutProvider(
+            graph_layout=self.pos)
+        # Attach the prepared renderer to the plot so it can be shown
+        self.plot.renderers.append(graph_renderer)
+
+    def _get_random_colors(self, num_colors=None):
+        colors = []
+        num_colors = num_colors or len(self.graph.vertices)
+        for _ in range(num_colors):
+            color = '#'+''.join([choice('0123456789ABCDEF') for j in range(6)])
+            colors.append(color)
+        return colors
+
```

Comment:
Consider challenging yourself more and hit some stretch goals. I'd like to see what you come up with! 👍